testing/fpgaconf/test_fpgaconf_c.cpp

```diff
@@ -0,0 +1,853 @@
+  print_err("msg", FPGA_OK);
+}
+
+/**
+ * @test       print_msg
+ * @brief      Test: print_msg
+ * @details    print_msg sends the given string to stdout<br>
+ *             if the given verbosity is less than or equal config.verbosity.<br>
+ */
+TEST_P(fpgaconf_c_p, print_msg) {
+  print_msg(0, "msg");
+}
+
+/**
+ * @test       read_bits_err0
+ * @brief      Test: read_bitstream
+ * @details    When given a NULL filename,<br>
+ *             read_bitstream returns a negative value.<br>
+ */
+TEST_P(fpgaconf_c_p, read_bits_err0) {
+  struct bitstream_info info;
+  EXPECT_LT(read_bitstream(nullptr, &info), 0);
+}
+
+/**
+ * @test       read_bits_err1
+ * @brief      Test: read_bitstream
+ * @details    When the filename is not found,<br>
+ *             read_bitstream returns a negative value.<br>
+ */
+TEST_P(fpgaconf_c_p, read_bits_err1) {
+  struct bitstream_info info;
+  EXPECT_LT(read_bitstream((char *) "/doesnt/exist", &info), 0);
+}
+
+/**
+ * @test       read_bits_err2
+ * @brief      Test: read_bitstream
+ * @details    When malloc fails,<br>
+ *             read_bitstream returns a negative value.<br>
+ */
+TEST_P(fpgaconf_c_p, read_bits_err2) {
+  struct bitstream_info info;
+  system_->invalidate_malloc(0, "read_bitstream");
+  EXPECT_LT(read_bitstream(tmp_gbs_, &info), 0);
+}
+
+/**
+ * @test       read_bits
+ * @brief      Test: read_bitstream
+ * @details    When the parameters to read_bitstream are valid,<br>
+ *             the function loads the given gbs file into the bitstream_info.<br>
+ */
+TEST_P(fpgaconf_c_p, read_bits) {
+  struct bitstream_info info;
+  EXPECT_EQ(read_bitstream(tmp_gbs_, &info), 0);
+  free(info.data);
+}
+
+/**
+ * @test       parse_args0
+ * @brief      Test: parse_args
+ * @details    When given an invalid command option,<br>
+ *             parse_args returns a negative value.<br>
+ */
+TEST_P(fpgaconf_c_p, parse_args0) {
+  char zero[20];
+  char one[20];
+  strcpy(zero, "fpgaconf");
+  strcpy(one, "-Y");
+
+  char *argv[] = { zero, one };
+
+  EXPECT_LT(parse_args(2, argv), 0);
+}
+
+/**
+ * @test       parse_args1
+ * @brief      Test: parse_args
+ * @details    When given valid command options,<br>
+ *             parse_args populates the global config struct<br>
+ *             and the fn returns 0.<br>
+ */
+TEST_P(fpgaconf_c_p, parse_args1) {
+  char zero[20];
+  char one[20];
+  char two[20];
+  char three[20];
+  char four[20];
+  char five[20];
+  char six[20];
+  char seven[20];
+  char eight[20];
+  char nine[20];
+  char ten[20];
+  char eleven[20];
+  char twelve[20];
+  char thirteen[20];
+  char fourteen[20];
+  strcpy(zero, "fpgaconf");
+  strcpy(one, "-v");
+  strcpy(two, "-n");
+  strcpy(three, "--force");
+  strcpy(four, "-B");
+  strcpy(five, "0x5e");
+  strcpy(six, "-D");
+  strcpy(seven, "0xab");
+  strcpy(eight, "-F");
+  strcpy(nine, "3");
+  strcpy(ten, "-S");
+  strcpy(eleven, "2");
+  strcpy(twelve, "-A");
+  strcpy(thirteen, "-I");
+  strcpy(fourteen, "file.gbs");
```

Comment:
Would this be easier to read if it were more like:
```C++
std::array<char*, 15> argv = { "fpgaconf", "-v", "-n", "--force",
                                "-B", "0x5e" /*, ...*/ };
```

Reply:
You would get warnings/errors about mixing const char * and char *.  That's why I coded it this way.  Note that argv is char *argv[].

Reply:
And so is the "std::array<char*, 15>" 😈 
You you just have to change the function call to use `argv.data()`

Reply:
But in your example, the string constants like "fpgaconf" are const char *, but the definition of your array is char *.  That's essentially the issue. :)

Reply:
Ok, you're right... boo! :)

Reply:
No harm done in questioning..  Glad to have convinced you. :)